Project 1 
Zack Owens
DATE SUBMITTED FIX THIS _________________________

# 1.1

In [4]:
eps32 = eps(0f0)

1.0f-45

In [5]:
eps64 = eps(0.0)

5.0e-324

Meanings of these values:
The method esp() according to the julia documentation is the distance between 1.0 and the next largest floating-point value of Float64. [Source](https://docs.julialang.org/en/v1/manual/integers-and-floating-point-numbers/) 

In [11]:
typeof(0f0)

(Float32, Inf32)

In [7]:
typeof(0.0)

Float64

The reason that 0f0 has a larger value is because the values is of type float32. This esp value the smaller the value the more precise value that can be stored in the type. 
The reason that 0.0 is a smaller values is due to its type being a float64 which has more bytes to store the float values. This gives teh float64 more percision than a float32 allowing smaller values to be stored in smaller increments.

# 1.2

In [16]:
v = fill(eps32,2_000_000_000)

2000000000-element Vector{Float32}:
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 ⋮
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45

In [ ]:
function sum_values_does_not_work(v)
    sum = 0.0
    for i in v
        sum += i
    end
    return sum
end

In [22]:
function sum_values(v)
    sum = 0.0
    for i in v
        sum += i
    end
    return sum
end

sum_values (generic function with 1 method)

In [23]:
@time sum_values(v)

  2.419582 seconds (7.53 k allocations: 510.795 KiB, 0.65% compilation time)


2.802596928649634e-36

In [28]:
@time foldr(+,v)

  2.612031 seconds (4 allocations: 64 bytes)


2.3509887f-38

In [27]:
@time foldr(+,v,init=0.0)

  2.552371 seconds (33.84 k allocations: 2.340 MiB, 1.89% compilation time)


2.802596928649634e-36

In [20]:
@time reduce(+,v)

  0.633802 seconds (108.92 k allocations: 7.286 MiB, 9.50% compilation time)


2.802597f-36

In [21]:
@time sum(v)

  0.584358 seconds (33.30 k allocations: 2.195 MiB, 5.93% compilation time)


2.802597f-36

The values above can be summarized by looking at the speeds and number of allocations for each function. The sum function performed the best with a correct solution. This function only took 0.584358 seconds with only 2.195 Megabytes of memory allocation. The next fastest function was the reduce function which took  0.633802 seconds with only 7.286 Megabytes of allocation. This method also created the correct answer. The foldr method was more than 4 times slower than the reduce and sum functions and performed the worst. This method took 2.552371 with 2.340 Megabytes of allocations. This method also did not create the correct answer unless the init paramters was specified. The for loop function performed the second worst out of all the fuctions and only returned the correct answer when sum was intitialized at value 0.0. This function completed in 2.419582 seconds with less allocations than reduce and sum.

The for loop and foldr did not create the correct answers unless the inital values specifed was a float32 instead julia assuming the answer is a float16 (a 0 or 0f0).